File: ml_test_vectors/stateless/flatten/implementation.py
from functools import reduce
from typing import Optional

import numpy as np

from ml_test_vectors.core import Tensor, Vector


def forward(inp: Tensor) -> Vector:
    def reducer(flattened: Vector, t: Tensor) -> Vector:
        return flattened + forward(t)

    initial_vector: Vector = []
    return [inp] if isinstance(inp, float) else reduce(reducer, inp, initial_vector)


def backward(inp: Tensor, _output: Optional[Tensor]) -> Tensor:
    return np.ones_like(inp).tolist()

File: ml_test_vectors/stateless/flatten/implementation.ts
import type {
	StatelessDerivativeFunction,
	StatelessFunction,
	Tensor,
	Vector,
} from "../../core";
import { isScalar, onesLike } from "../../utils";

export const forward: StatelessFunction<Tensor, Vector> = (input) =>
	isScalar(input)
		? [input]
		: input.reduce<Vector>((flattened, t) => [...flattened, ...forward(t)], []);

export const backward: StatelessDerivativeFunction = (input) => onesLike(input);
